### PR TITLE
[ESQL] Move optimizer rules out of LogicalPlanOptimizer

### PR DESCRIPTION
**Pull request by not-napoleon**
This moves all of the actual rule classes out of `LogicalPlanOptimizer`, where they were static sub-classes, into top level classes in the `optimizer.rules` package.  This improves the readability of `LogicalPlanOptimizer`, and of the individual rules, while also reducing the chances of merge conflicts while touching these rules (something that may happen a lot as we try to move things out of core).

### CODE REVIEW

**Comment by not-napoleon** [comment]
Apologies in advance for this being a large delta of lines.  It's all automated refactoring, just moving code around.  Hopefully, the separation of concerns here means that future things touching this code can do so in parallel better, and thus support smaller PRs.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 1]
```diff
@@ -1746,114 +302,4 @@ public final LogicalPlan apply(LogicalPlan plan, P context) {
...
         protected abstract LogicalPlan rule(SubPlan plan, P context);
     }
-
-    /**
-     * Normalize aggregation functions by:
-     * 1. replaces reference to field attributes with their source
-     * 2. in case of Count, aligns the various forms (Count(1), Count(0), Count(), Count(*)) to Count(*)
-     */
-    // TODO still needed?
-    static class NormalizeAggregate extends Rule<LogicalPlan, LogicalPlan> {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1756]: NB: This rule was deleted, not pulled up to a top level class.  There were no usages of it.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/FoldNull.java [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 import org.elasticsearch.xpack.esql.core.expression.Expression;
 
-public class FoldNull extends LogicalPlanOptimizer.FoldNull {
+public class FoldNull extends org.elasticsearch.xpack.esql.optimizer.rules.FoldNull {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/FoldNull.java:12]: Why does this class exist?  Really feels like we can delete it and just reference the superclass in all cases.

Comment (luigidellaquila) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/FoldNull.java:12]: Maybe it was for visibility issues in the past (the base class was not public in QL...?), but as long as it compiles, I guess we can get rid of this one now.

**Review by luigidellaquila — APPROVED** [review]
LGTM, thanks @not-napoleon!

**Review by astefan — COMMENTED** [review]
Please, make all the extracted rules `final`.
Also, note that not using the static class approach from the optimizer anymore we are not necessarily doing the same with non-static classes in `*.rules` package.

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**